From: SeongJae Park <sj@kernel.org>
Date: Thu, 19 Jun 2025 10:40:19 -0700
Subject: [PATCH] mm/damon/sysfs-schemes: free
 damon_sysfs_scheme_filter->memcg_path on write

memcg_path_store() sets filter->memcg_path with newly allocated memory
buffer for the path, without deallocating the previously set memory
buffer.  As a result, users can leak kernel memory by continuously
writing a data to memcg_path DAMOS sysfs file.  Fix the leak by
deallocating the previously set memory buffer.

Fixes: 7ee161f18b5d ("mm/damon/sysfs-schemes: implement filter directory")
Cc: stable@vger.kernel.org	# 6.3.x
Signed-off-by: SeongJae Park <sj@kernel.org>
---
 mm/damon/sysfs-schemes.c | 1 +
 1 file changed, 1 insertion(+)

diff --git a/mm/damon/sysfs-schemes.c b/mm/damon/sysfs-schemes.c
index 5a450e362af0..4805761d9127 100644
--- a/mm/damon/sysfs-schemes.c
+++ b/mm/damon/sysfs-schemes.c
@@ -511,6 +511,7 @@ static ssize_t memcg_path_store(struct kobject *kobj,
 		return -ENOMEM;
 
 	strscpy(path, buf, count + 1);
+	kfree(filter->memcg_path);
 	filter->memcg_path = path;
 	return count;
 }
-- 
2.39.5

